docs/architecture/adr-070-unordered-account.md

```diff
@@ -2,214 +2,134 @@
-  // blockCh defines a channel to receive newly committed block time
-  blockCh chan time.Time
+import (
+  "fmt"
+  "time"
 
-  mu sync.RWMutex
-	// txHashes defines a map from tx hash -> TTL value, which is used for duplicate
-	// checking and replay protection, as well as purging the map when the TTL is
-	// expired.
-	txHashes map[TxHash]time.Time
-}
+  sdk "github.com/cosmos/cosmos-sdk/types"
 
-func NewUnorderedTxManager() *UnorderedTxManager {
-  m := &UnorderedTxManager{
-		blockCh:  make(chan time.Time, 16),
-		txHashes: make(map[TxHash]time.Time),
-  }
+  "cosmossdk.io/collections"
+  "cosmossdk.io/core/store"
+)
 
- return m
-}
+const (
+  // DefaultMaxTimeoutDuration defines the default maximum duration an unordered transaction
+  // can set.
+  DefaultMaxTimeoutDuration = time.Minute * 10
+)
 
-func (m *UnorderedTxManager) Start() {
-  go m.purgeLoop()
-}
+type UnorderedSequence string
 
-func (m *UnorderedTxManager) Close() error {
-  close(m.blockCh)
-  m.blockCh = nil
-  return nil
+func NewUnorderedSequence(addr string, timestamp uint64) UnorderedSequence {
+  return UnorderedSequence(fmt.Sprintf("%d/%s", timestamp, addr))
 }
 
-func (m *UnorderedTxManager) Contains(hash TxHash)  bool{
-  m.mu.RLock()
-  defer m.mu.RUnlock()
-
-  _, ok := m.txHashes[hash]
-  return ok
+// Manager contains the tx hash dictionary for duplicates checking, and expire
+// them when block production progresses.
+type Manager struct {
+  kvStore store.KVStoreService
 }
 
-func (m *UnorderedTxManager) Size() int {
-  m.mu.RLock()
-  defer m.mu.RUnlock()
+func NewManager(kvStore store.KVStoreService) *Manager {
+  m := &Manager{
+    kvStore: kvStore,
+  }
 
-  return len(m.txHashes)
+  return m
 }
 
-func (m *UnorderedTxManager) Add(hash TxHash, expire time.Time) {
-  m.mu.Lock()
-  defer m.mu.Unlock()
-
-  m.txHashes[hash] = expire
+func (m *Manager) Contains(ctx sdk.Context, sender string, timestamp uint64) (bool, error) {
+  return m.kvStore.OpenKVStore(ctx).Has([]byte(NewUnorderedSequence(sender, timestamp)))
 }
 
-// OnNewBlock send the latest block time to the background purge loop, which
-// should be called in ABCI Commit event.
-func (m *UnorderedTxManager) OnNewBlock(blockTime time.Time) {
-  m.blockCh <- blockTime
+func (m *Manager) Add(ctx sdk.Context, sender string, timestamp uint64) error {
+  return m.kvStore.OpenKVStore(ctx).Set([]byte(NewUnorderedSequence(sender, timestamp)), []byte(byte(0x0)))
 }
 
-// expiredTxs returns expired tx hashes based on the provided block time.
-func (m *UnorderedTxManager) expiredTxs(blockTime time.Time) []TxHash {
-  m.mu.RLock()
-  defer m.mu.RUnlock()
-
-  var result []TxHash
-  for txHash, expire := range m.txHashes {
-    if blockTime.After(expire) {
-      result = append(result, txHash)
-    }
+func (m *Manager) RemoveExpired(ctx sdk.Context) error {
+  kvStore := m.kvStore.OpenKVStore(ctx)
+  it, err := kvStore.Iterator(nil, []byte(NewUnorderedSequence("", uint64(ctx.BlockTime().UnixNano()))))
```

Comment:
So this runs from "the beginning to the blocktime"?

Reply:
yes